### PR TITLE
Added "Deploy to Heroku with PostgreSQL" chapter including way to automatically run pending migrations on Heroku

### PR DESCRIPTION
**Pull request by mattwelke**
Hope my first contribution with new content looks good! I tried to follow the style and tone of the book as best as I could (friendly, conversational, but still short and to the point).

Every step I instruct the reader to do in the book is assuming they've completed the book so far at least to the point where they add the `TodoItem` model. I also tested each step as I went to make sure it worked properly, and it worked on my machine (Windows 10, PostgreSQL 10.2 via Docker).

I was able to get my own fork of the book hosted (https://welkie.gitbooks.io/littleaspnetcorebookfork/content/chapters/alternative-databases/postgresql.html) if you want to take a look and see how this would render if the pull request is accepted. I welcome your input on the way the code examples are laid out, etc.

### CODE REVIEW

**Comment by nbarbettini** [comment]
Thanks @welkie! FYI, I'm currently super busy helping organize a conference in SF (https://iterateconf.io), so I won't be able to take a look right away. I'm excited to get contributions though! 😄

**Comment by welkie** [comment]
@nbarbettini Sounds good. If you don't mind me bugging you for some advice though in the mean time, I was going to contribute another chapter or fix a few up according to the issues that we've got logged right now. But I'm new to submitting pull requests to other projects, so I'm not sure how to go about doing that. If I added to this pull request, it wouldn't be ideal because accepting all changes would be all or nothing. But if I do new branches for each addition I do, with individual pull requests, then the 2nd 3rd etc wouldn't be able to be merged after the 1st is merged (I'd have to manually update the branches for them with the latest code now that the 1st was merged). Unless my understanding of that isn't good? I'd appreciate some advice if you know the best way to do contributions like that.

**Comment by sadqiang** [comment]
Your link goes to a 404 page. Please kindly check it again. Thank you!

**Comment by welkie** [comment]
@sadqiang GitBook unfortunately disabled my account after I created that book. They classified it as spam. I think they got confused assuming I was intending to commit copyright infringement or something. Unfortunately, we can't see my changes rendered unless we know how to simulate GitBook locally or the pull request is accepted. You can see the changes by checking the "Files changed" in this pull request though.

**Comment by nbarbettini** [comment]
@welkie Good question - I'd definitely prefer to keep pull requests separate instead of adding to one giant PR.

If you create a new branch `other-changes` off `master` and make edits that don't depend on anything in this branch, then `other-changes` can be PRed and merged to `master` independently of this one. You won't have to wait for this PR to be merged.

Also, you can always merge or rebase the latest changes from `master` into a branch you are working on to keep it "fresh". My typical process for working on a branch is:

* Create branch off master: `git checkout -b other-changes`
* Edit or add content, push up to Github
* Open a PR against `master` when it's ready for review
* Make additional changes/updates
* If necessary, rebase against `master` so everything in the branch sits on top of the latest commits to `master`
* Merge via Github

**Comment by nbarbettini** [comment]
Hopefully I understood your question and scenario. If you need some help or just want to see how it works, shoot me an email and we can do a quick screenshare call.

**Comment by nbarbettini** [comment]
@welkie @sadqiang It is possible to use Gitbook locally: https://toolchain.gitbook.com/setup.html

That's super weird that Gitbook disabled your account. Other forks of the book have been hosted on Gitbook just fine. /shrug

**Comment by welkie** [comment]
@nbarbettini That's basically how I planned to do it. Separate branches each with PRs. I didn't realize you could submit PRs for merging those branches into the origin repo's master. That helps and should keep the work for the origin repo's maintainers limited. They would just need to accept the individual PRs which probably won't conflict. I think that clears it up for me. I'll leave you alone to prepare for Iterate. Good luck! :+1:

File: chapters/alternative-databases/README.md [hunk 1]
```diff
@@ -0,0 +1,2 @@
+# Alternative databases
+Now that you've learned how to create and deploy a basic ASP.NET Core application using SQLite, you can explore other databases you can use.
```
Comment (nbarbettini) [inline chapters/alternative-databases/README.md]: Let's add a note here explaining that this chapter is optional.

File: SUMMARY.md [hunk 2]
```diff
@@ -31,4 +31,6 @@
 * [Deploy the application](chapters/deploy-the-application/README.md)
     * [Deploy to Azure](chapters/deploy-the-application/deploy-to-azure.md)
     * [Deploy with Docker](chapters/deploy-the-application/deploy-with-docker.md)
+* [Alternative databases](chapters/alternative-databases/README.md)
```
Comment (nbarbettini) [inline SUMMARY.md]: I'm still trying to decide where this should go. I think it makes more sense before Deploy, but then the Deploy chapter might have to worry about which database people are using.

File: chapters/alternative-databases/postgresql.md [hunk 3]
```diff
@@ -0,0 +1,87 @@
+## PostgreSQL
+
+Using PostgreSQL for your application instead of SQLite will allow it to scale better and give you access to various PostgreSQL features like the [PostGIS extension](https://postgis.net/).
+
+Thanks to the layer of abstraction we get from using Entity Framework Core, modifying our application to use a different database engine is less complex than you would think. We can use our existing models, only having to change the Entity Framework Core provider we use. We will add the [Npgsql EF Core provider](http://www.npgsql.org/efcore/index.html).
```
Comment (nbarbettini) [inline chapters/alternative-databases/postgresql.md]: I tried to avoid our/we in the book and stick with "you" - can you change the phrasing in this file to match?

File: chapters/alternative-databases/postgresql.md [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+
+### Add provider Nuget package
```
Comment (nbarbettini) [inline chapters/alternative-databases/postgresql.md]: nit: NuGet

File: chapters/alternative-databases/postgresql.md [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+
+```
+dotnet ef database update
+```
+
+While it is true that the migrations we created in the project have already been applied, they were applied when the project was configured to use SQLite. Entity Framework Core knows that with this new database configuration, the migrations have not yet been applied, so it applies them to the PostgreSQL database, resulting in the tables being created. You can use a program like [pgAdmin](https://www.pgadmin.org/) to connect to the database and see the tables that were created:
+
+![pgAdmin showing the tables](pgadmin_aspnetcore_tables_created.png)
+
+At this point, you can start your application, visit `localhost:5000` in your browser, and create and manipulate `Todo` items just like before, except this time the data is being saved in the PostgreSQL database. Notice how everything works as before, including:
```
Comment (nbarbettini) [inline chapters/alternative-databases/postgresql.md]: Can you rephrase this to avoid saying "at this point" twice?

**Review by nbarbettini — CHANGES_REQUESTED** [review]

**Comment by nbarbettini** [comment]
Reviewed. I like it, thanks for the contribution! The tone fits the book well.

The only thing I'm still thinking through is 
* Where it should go chronologically. It currently feels a little weird to explore this after the application is deployed.
* What other databases could be discussed in this chapter - I like PostgreSQL but it's lonely! 😄

**Comment by welkie** [comment]
For sure. I'm thinking it's worth at least mentioning MongoDB, even though its provider is very beta right now. MySQL also has pretty good support.

I'll touch up the minor things from the review soon. In the mean time, if you have an idea for the best place chronologically for this topic, be sure to let me know.

File: chapters/alternative-databases/README.md (hunk 1 above)
Comment (mattwelke) [reply chapters/alternative-databases/README.md]: Added note indicating optional. This may change if chapter rearranged in future.

File: SUMMARY.md (hunk 2 above)
Comment (mattwelke) [reply SUMMARY.md]: Awaiting feedback on proposal in comment below.

**Comment by welkie** [comment]
I left the arrangement alone for now, but cleaned it up using your feedback. You can see the changes in https://github.com/nbarbettini/little-aspnetcore-book/pull/49/commits/817bbf6be05ea8e2a6e6dc0804a988454d9106f8.

File: chapters/alternative-databases/postgresql.md (hunk 3 above)
Comment (mattwelke) [reply chapters/alternative-databases/postgresql.md]: Changed to "you" throughout chapter.

File: chapters/alternative-databases/postgresql.md (hunk 4 above)
Comment (mattwelke) [reply chapters/alternative-databases/postgresql.md]: Changed to NuGet throughout chapter.

File: chapters/alternative-databases/postgresql.md (hunk 5 above)
Comment (mattwelke) [reply chapters/alternative-databases/postgresql.md]: Replaced all 'at this point' phrases with descriptive paragraph intros.

**Comment by nbarbettini** [comment]
@welkie I really like the idea of a Deploy to Heroku chapter! Would you be willing to write that?

This logical progression makes the most sense to me so far:

* Use a database (note that it's SQLite, good for development)
* ...
* Deploy to Azure
* * Use Azure SQL Server
* Deploy to Heroku
* * Use PostgreSQL
* Deploy with Docker (with SQLite, mention that Postgres works too)

**Comment by welkie** [comment]
That sounds good to me! I played around with the Heroku buildpack before and I noticed some rough edges. Not sure if it was me or the buildpack itself. I'm going to look more into it and if I find that it can be taught simply I'll make that new chapter and merge the Postgres like content in.

I'm prepping for exams right now, so that will be something I get started with later.

**Comment by welkie** [comment]
@nbarbettini I was able to get my Heroku deploy working.

[This thread](https://github.com/jincod/dotnetcore-buildpack/issues/33) details what I had to do to get around the inability to run migrations via the Heroky CLI right now (at this point I'm not sure if that's a bug with the buildpack or if that's just me being dumb).

I don't think it's that bad actually now that I've done it. I think it can be explained simply in a Heroku chapter. Let me know what you think of that thread and if you give me the go ahead, I'll work on the Heroku chapter shortly and integrate my PostgreSQL stuff into it.

**Comment by nbarbettini** [comment]
If the author of the buildpack doesn't get back to you soon, I think it makes sense to go ahead and doc the workaround. 👍

**Comment by nbarbettini** [comment]
FYI I'm releasing version 1.1 today and planning on releasing another large update around the time that ASP.NET Core 2.1 is final. The new stuff we've been discussing here fits in well with the next update I think. 👍

**Comment by welkie** [comment]
@nbarbettini I agree. I've been busy lately starting my first post grad job so this hasn't been on my radar. It will be soon though. :)

**Comment by mattwelke** [comment]
Wanted to provide an update here. I was coordinating with the maintainers behind the [.NET Core buildpack](https://github.com/jincod/dotnetcore-buildpack) to find out the best way to go about deploying .NET Core to Heroku. There ended up being some [disagreement](https://github.com/jincod/dotnetcore-buildpack/pull/47) on the best way to handle db migrations. There were two solutions found:

- Adding code to manually grab a DbContext instance during program startup and running pending migrations.
- Using a "Target" (`<Exec Command="dotnet ef database update" />`) to run the SDK during the deploy step to run pending migrations.

And another user put forth an idea, to use the official Heroku "Release" system to do migrations.

I was waiting for them to sort out what they think would be the best option but it hasn't gone anywhere and I'd like to finish the PR since I sent it in February and I think it would be useful for users. I'd welcome some input here and then I can finish the chapter with how we'll instruct our readers to deploy on Heroku.

**Comment by mattwelke** [comment]
@nbarbettini 
Rebased and redid the chapter to be a new "Deploy to Heroku with PostgreSQL" chapter. The debate on the best way to handle Heroku side migrations was never solved, so I decided to go with the solution where the source code is changed to run pending migrations automatically as the application starts up if it detects that it's running on Heroku. You can take a look at the commits to see this. Despite 2.1 having been released, which includes some slight differences in how the `Program.cs` code would look, I kept this based on 2.0, since we can tackle the 2.1 update later on.

**Comment by mattwelke** [comment]
@nbarbettini Was wondering if you've had a chance to look at this? Also, one thing I was thinking was that this might be too much for this book, since its goal is to be focused on just introducing ASP.NET Core to people, and it already covers one deployment method. Curious what your thoughts are.

**Review by gaKoltai — APPROVED** [review]
